### PR TITLE
feat: past and overlapping ooo

### LINKED ISSUE #17368: [CAL-4631] allow OOO of the past and overlapping
for tracking purposes it would be dope to add OOO retroactively. currently its blocked:

image

it would also be nice if OOO could overlap, there is no reason someone cant be OOO for sick day and then when that ends be OOO for traveling or so

right now we throw an error for overlapping:

image

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-4631](https://linear.app/calcom/issue/CAL-4631/allow-ooo-of-the-past-and-overlapping)</sub>

### LINKED ISSUE #17772: [CAL-4769] currently cant have overlapping OOO
ijright now in [https://ooo.new](https://ooo.new) its a strict requirement that OOO cannot overlap. it throws an error

however, i think we should allow it

for example, if i go OOO 1 to 12 for "travel", i should still afterwards be able to go OOO on the 6th for "vacation"

the last entry will always overwrite, in this example it would show the travel emoji from 1-5 and 6 would be the vacation emoji and 7-12th would be travel again

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-4769](https://linear.app/calcom/issue/CAL-4769/currently-cant-have-overlapping-ooo)</sub>

### PR DESCRIPTION
**Pull request by Amit91848**
## What does this PR do?

- Fixes #17368 
- Fixes CAL-4631 
/claim #17368
- Fixes #17772

[Screencast from 2024-10-28 21-26-38.webm](https://github.com/user-attachments/assets/a9d8c437-995b-4204-a8e8-0cc2bc36cf3e)

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

### CODE REVIEW

File: packages/ui/components/form/date-range-picker/Calendar.tsx [hunk 1]
```diff
@@ -9,19 +9,22 @@ import { classNames as cn } from "@calcom/lib";
...
+  disablePastDates?: boolean;
+};
 
 function Calendar({
   className,
   classNames,
   fromDate,
   toDate,
   showOutsideDays = true,
+  disablePastDates = true,
```
Comment (emrysal) [inline packages/ui/components/form/date-range-picker/Calendar.tsx:22]: would prefer this to be a minDate option (set by default to new Date()) - and nullable to allow all dates enabled.

**Review by emrysal — DISMISSED** [review]
https://github.com/calcom/cal.com/pull/17373/files#r1822623987

File: packages/ui/components/form/date-range-picker/Calendar.tsx [hunk 2]
```diff
@@ -21,7 +21,7 @@ function Calendar({
 }: CalendarProps) {
   return (
     <DayPicker
-      fromDate={fromDate || new Date()}
+      fromDate={fromDate}
```
Comment (Amit91848) [inline packages/ui/components/form/date-range-picker/Calendar.tsx:24]: Handled in parent component

File: packages/ui/components/form/date-range-picker/DateRangePicker.tsx [hunk 3]
```diff
@@ -63,7 +64,8 @@ export function DatePickerWithRange({
           sideOffset={4}>
           <Calendar
             initialFocus
-            fromDate={minDate}
+            //When explicitly null, we want past dates to be shown as well, otherwise show only dates passed or from current date
```
Comment (Amit91848) [inline packages/ui/components/form/date-range-picker/DateRangePicker.tsx:67]: Added comment because this is confusing

Comment (Praashh) [reply packages/ui/components/form/date-range-picker/DateRangePicker.tsx:67]: yeah it seems you are handling null Date and minDate conditions inside the `DateRangePicker`, which is good.

**Review by Praashh — APPROVED** [review]
changes looks correct to me, tested in local as well. Thank you @Amit91848

**Review by anikdhabal — DISMISSED** [review]
Pls add test to verify this

**Comment by Amit91848** [comment]
https://github.com/user-attachments/assets/04032e0b-9c2a-4480-b701-24bb715d7282

Is something up with ooo page? I cannot load it and it straight up crashes the server without any error message. I have the latest pull from upstream

**Comment by Praashh** [comment]
> Recording.2024-11-18.132444.mp4 
> Is something up with ooo page? I cannot load it and it straight up crashes the server without any error message. I have the latest pull from upstream

nope, working for me

**Review by Praashh — CHANGES_REQUESTED** [review]
All good, can you please add a test case for overlapping OOOs?

**Comment by Amit91848** [comment]
Added tests, sorry for the delay.

Pretty sure there is a memory leak in ooo page. Also maybe in other sub pages under /my-account

Screenshot 2024-11-20 205438

Screenshot 2024-11-20 205634

**Comment by Praashh** [comment]
for me it's working fine waiting for @anikdhabal to check this.

**Review by Praashh — CHANGES_REQUESTED** [review]
image


I notice that I don't think we need duplicate OOO, we only need overlapping OOOs.

File: packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts [hunk 4]
```diff
@@ -73,51 +68,6 @@ export const outOfOfficeCreateOrUpdate = async ({ ctx, input }: TBookingRedirect
     toUserId = user?.id;
   }
 
-  // Validate if OOO entry for these dates already exists
```
Comment (Praashh) [inline packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:76]: Here we can add a check for duplicate OOOs

**Comment by Amit91848** [comment]
> image
> 
> I notice that I don't think we need duplicate OOO, we only need overlapping OOOs.

Nice catch, made changes and updated tests

**Review by Praashh — APPROVED** [review]
LGTM!!

**Pull request merged**